### DISCUSSION TITLE
How to integrate lexical for submitting Editor as input

### DISCUSSION THREAD

**Discussion by PunkFleet** [facebook/lexical#3720] [Q&A] [answered]
When I try to make a blog, I can't pass the editor in the form. I found this:

https://stackoverflow.com/questions/63122422/draftjs-react-hook-form-submitting-editor-as-input

but it seems that LexicalRichTextEditor does not have such a tag to pass. Can anyone help me?
```
type LexicalEditorProps = {
    //config: Parameters<typeof LexicalComposer>["0"]["initialConfig"];
    content: any;
};

export default function MyEditor(props: LexicalEditorProps) {
    const [ editor ] = useLexicalComposerContext();
    const editorStateRef = useRef();
    const [saveContent, setSaveContent] = useState('');
    const editorConfig: any = {
        // The editor theme
        theme: EditorTheme,
        // Handling of errors during update
        onError(error: any) {
            throw error;
        },
        editorState: props.content,
        // Any custom nodes go here
        nodes: [
            HeadingNode,
            ListNode,
            ListItemNode,
            QuoteNode,
            CodeNode,
            CodeHighlightNode,
            TableNode,
            TableCellNode,
            TableRowNode,
            AutoLinkNode,
            LinkNode
        ]
    };

    useEffect(()=>{
        if(editorStateRef.current){
            setSaveContent(JSON.stringify(editorStateRef.current));
        }
        editor.update(()=>{
            const root = $getRoot();
            const selection = $getSelection();
            const paragraphNode = $createParagraphNode();
            const textNode = $createTextNode(saveContent);

            paragraphNode.append(textNode);
            root.append(paragraphNode);

        });
    },[saveContent]);
    return (
        <LexicalComposer initialConfig={editorConfig}>
            <div className="editor-container">
                <ToolbarPlugin />
                <div className="editor-inner">
                    <RichTextPlugin
                        contentEditable={<ContentEditable className="editor-input" />}
                        placeholder={<Placeholder />}
                        ErrorBoundary={LexicalErrorBoundary}
                    />
                    <OnChangePlugin onChange={(editorState:any) => editorStateRef.current = editorState} />
                    <HistoryPlugin />
                    <AutoFocusPlugin />
                    <CodeHighlightPlugin />
                    <ListPlugin />
                    <LinkPlugin />
                    <AutoLinkPlugin />
                    <ListMaxIndentLevelPlugin maxDepth={7} />
                    <MarkdownShortcutPlugin transformers={TRANSFORMERS} />
                </div>
            </div>
        </LexicalComposer>
    );
}

export function MyForm(){
 const {register, handleSubmit, control, formState: {errors}} = useForm();
    const onSubmit = ( data:any) => {      
        console.log(data);
    };
 return (
   <form onSubmit={handleSubmit(onSubmit)}>
                <Stack spacing={2}>                                      
                    <Card>
                        <Controller control={control} name="content" render={()=> (
                            <MyEditor content={dataSet.content} />
                        )} />
                    </Card>
                    <Box>
                        <Button variant="contained" type="submit">Save</Button>
                    </Box>
                </Stack>
            </form>
 );
}
```

**Comment by moy2010** [comment]
What do you need react-hook-form for? It's meant to be used for form validation, and I can't think of any scenario where you want both a rich text editor and form validation.

**Reply by PunkFleet** [reply]
![image](https://user-images.githubusercontent.com/86408260/212469074-08048b65-862f-431c-ae41-42bdddbc08aa.png)
I use it for writing blogs and I don't have much programming experience, isn't that how other people use their blogs? I'm guessing that such a scenario should be pretty generic.

**Reply by PunkFleet** [reply]
I guess the blog application should submit category, tag, content, title, date... as each form field submitted to the database, so when I wrote it, I found I couldn't pass the EDITOR content to the form

**Reply by moy2010** [reply]
Imagine that you only want users to enter post titles up to 45 characters long. That's what you would use react-hook-form, so that you can show a nice message when a user get above that limit.

For the text editor part, unless your use case requires it, there shouldn't be any validation at all.

**Reply by PunkFleet** [reply]
![image](https://user-images.githubusercontent.com/86408260/212519507-63576713-3167-4f3f-a170-e8b7fa926961.png)
When I simply add the editor to the form and don't do anything with it, I can't get the editor fields

**Comment by PunkFleet** [comment]
Yeah, yeah, that's exactly what I'm aiming for. Sorry I am indeed a newbie. Can I do this: apply react-hook-form to only some fields, other fields don't need it. Can you give me some hints? How to get or pass the lexical value in the form. This code does not work, i got hint:
```
LexicalComposerContext.useLexicalComposerContext: cannot find a LexicalComposerContext
```

**Reply by trueadm** [reply]
You can’t pass Lexical into a form as it’s not a form field. Instead, you’ll need to extract the contact via Lexical’s APIs upon form submit. In order to get a reference to the Lexical editor, you’ll need to create a Lexical plugin that takes a React ref as a prop, and in an effect, assigns the ref.current = editor

**Reply by PunkFleet** [reply]
Thank you @trueadm 
I tried to use forwardRef, but I couldn't find how to add the ref attribute to the lexical. Here is the code I changed, it doesn't report an error but it doesn't seem to have any effect either, can you help me make some changes?
```
export const MyEditor = forwardRef((props: any, ref) => {
    const editorStateRef = useRef();
    const [saveContent, setSaveContent] = useState('');
    const editorConfig: any = {
        // The editor theme
        theme: EditorTheme,
        // Handling of errors during update
        onError(error: any) {
            throw error;
        },
        editorState: props.content,
        // Any custom nodes go here
        nodes: [
            HeadingNode,
            ListNode,
            ListItemNode,
            QuoteNode,
            CodeNode,
            CodeHighlightNode,
            TableNode,
            TableCellNode,
            TableRowNode,
            AutoLinkNode,
            LinkNode
        ]
    };

    useEffect(()=>{
        if(editorStateRef.current){
            setSaveContent(JSON.stringify(editorStateRef.current));
        }
        
    },[saveContent]);
    const editorContent = (editorState:any) => {
        ref = editorState
    };
    return (
        <LexicalComposer initialConfig={editorConfig}>
            <div className="editor-container">
                <ToolbarPlugin />
                <div className="editor-inner">
                    <RichTextPlugin
                        contentEditable={<ContentEditable className="editor-input" />}
                        placeholder={<Placeholder />}
                        ErrorBoundary={LexicalErrorBoundary}
                    />
                    <OnChangePlugin onChange={editorContent} />
                    <HistoryPlugin />
                    <AutoFocusPlugin />
                    <CodeHighlightPlugin />
                    <ListPlugin />
                    <LinkPlugin />
                    <AutoLinkPlugin />
                    <ListMaxIndentLevelPlugin maxDepth={7} />
                    <MarkdownShortcutPlugin transformers={TRANSFORMERS} />
                </div>
            </div>
        </LexicalComposer>
    );
});

export function MyForm(){
 const {register, handleSubmit, control, formState: {errors}} = useForm();
    const onSubmit = ( data:any) => {      
        console.log(data);
        console.log(formRef);
    };

   const formRef = useRef();
 
 return (
   <form onSubmit={handleSubmit(onSubmit)}>
                <Stack spacing={2}> 
                    //...other fields                                     
                    <Card>                        
                            <MyEditor ref={formRef} />                     
                    </Card>
                    <Box>
                        <Button variant="contained" type="submit">Save</Button>
                    </Box>
                </Stack>
            </form>
 );
}
```

**Reply by trueadm** [reply]
You’re not using the forwarded ref though? You’re using the useRef? You need do useRef outside the component and pass that in a as a ref, then reference the forwarded ref.

**Reply by PunkFleet** [reply]
Oh, I see, I used them in the wrong place, I should have switched them🤣

**Reply by trueadm** [reply]
```js
useEffect(()=>{
  ref.current = editor;

  return () => {
    ref.current = null;
  }
},[saveContent]);
```

**Reply by trueadm** [reply]
I added an example above how what I meant. You can then access the editor from the `ref` that you pass into your main editor component and use the editor APIs to extract the content as you want.

```js
  // Outside
  if (ref.current !== null 
    const latestEditorState = ref.current.getEditorState()
    // Do something? i.e. get the text of the editor
    const textContent = latestEditorState.read(() => $getRoot.getTextContent());
  }
``

**Reply by PunkFleet** [reply]
I seem to understand the logic of it, let me try to solve this problem, BTW, will this feature be added in the next update version? I think maybe this feature will make lexical truly out of the box, or at least will make newbies like me switch from quill to here🥳

**Reply by PunkFleet** [reply]
> I added an example above how what I meant. You can then access the editor from the `ref` that you pass into your main editor component and use the editor APIs to extract the content as you want.
> 
> ```js
>   // Outside
>   if (ref.current !== null 
>     const latestEditorState = ref.current.getEditorState()
>     // Do something? i.e. get the text of the editor
>     const textContent = latestEditorState.read(() => $getRoot.getTextContent());
>   }
> ``
> ```

https://codesandbox.io/s/heuristic-wildflower-b1l39f?file=/src/MyEditor.tsx
Hi Boss, Is this logic correct? i update it to codesandbox.

**Reply by Ou7law007** [reply]
> LexicalComposerContext.useLexicalComposerContext: cannot find a LexicalComposerContext

I'm getting this same error. Can you tell me how you got and how you fixed it?

**Reply by PunkFleet** [reply]
You're using useRef in the wrong place i guess.
Just calling useRef and direct using 'editorRef.current.getEditorState()', Please checking this:
https://stackoverflow.com/questions/75065990/how-to-integrate-lexical-with-react-hook-form-for-submitting-editor-as-input/75141769#75141769

**Comment by PunkFleet** [comment, marked as answer]
·············update····················
There is my question has a right solutions:
https://stackoverflow.com/questions/75065990/how-to-integrate-lexical-with-react-hook-form-for-submitting-editor-as-input/75141769#75141769
···········································
@trueadm Really, really appreciate your help, Thank you. And thanks to another great friend @abelsj60 for providing the solution. Hope any inspiration or help for future friends.
https://codesandbox.io/s/purple-water-xf50bi?file=/src/App.tsx

```
export default function App() {
  const schema = yup
    .object({
      title: yup.string().required(),
      category: yup.string().required(),
      tags: yup.array().required()
    })
    .required();
  const { register, handleSubmit } = useForm({
    resolver: yupResolver(schema)
  });
  const onSubmit = (data: any) => {
    console.log(data);
  };
  const editorRef: any = useRef();
  // ADDED THIS:
  if (editorRef.current !== undefined) {
    if (editorRef.current !== null) {
      const latestEditorState = editorRef.current.getEditorState();
      const textContent = latestEditorState.read(() =>
        //You could change getTextContent() for your purpose
        $getRoot().getTextContent()
      );
      console.log(textContent);
    }
  }
  return (
    <div className="App">
      <form onSubmit={handleSubmit(onSubmit)}>
        <Stack spacing={2}>
          <input {...register("title")} type="text" placeholder="Title" />
          <input {...register("category")} type="text" placeholder="Category" />
          <input type="" placeholder="Tags" />
          <select {...register("tags")} id="tags" multiple>
            <option value="nginx">nginx</option>
            <option value="java">java</option>
            <option value="react">react</option>
            <option value="mui">mui</option>
          </select>
          <Card elevation={3}>
            <MyEditor ref={editorRef} />
          </Card>
          <button type="submit">Save</button>
        </Stack>
      </form>
    </div>
  );
}
```
MyEditor.tsx
```
function Placeholder() {
  return <div className="editor-placeholder">Enter some rich text...</div>;
}

const editorConfig: any = {
  // The editor theme
  theme: EditorTheme,
  // Handling of errors during update
  onError(error: any) {
    throw error;
  },
  // Any custom nodes go here
  nodes: [
    HeadingNode,
    ListNode,
    ListItemNode,
    QuoteNode,
    CodeNode,
    CodeHighlightNode,
    TableNode,
    TableCellNode,
    TableRowNode,
    AutoLinkNode,
    LinkNode
  ]
};

// ADDED THIS:
const EditorCapturePlugin = React.forwardRef((props: any, ref: any) => {
  const [editor] = useLexicalComposerContext();
  useEffect(() => {
    ref.current = editor;
    return () => {
      ref.current = null;
    };
  }, [editor, ref]);

  return null;
});

export const MyEditor = React.forwardRef((props: any, ref: any) => {
  return (
    <LexicalComposer initialConfig={editorConfig}>
      <div className="editor-container">
        <ToolbarPlugin />
        <div className="editor-inner">
          <RichTextPlugin
            contentEditable={<ContentEditable className="editor-input" />}
            placeholder={<Placeholder />}
            ErrorBoundary={LexicalErrorBoundary}
          />
          {/* ADDED THIS: */}
          <EditorCapturePlugin ref={ref} />
          <HistoryPlugin />
          <AutoFocusPlugin />
          <CodeHighlightPlugin />
          <ListPlugin />
          <LinkPlugin />
          <AutoLinkPlugin />
          <ListMaxIndentLevelPlugin maxDepth={7} />
          <MarkdownShortcutPlugin transformers={TRANSFORMERS} />
        </div>
      </div>
    </LexicalComposer>
  );
});
```

**Reply by PunkFleet** [reply]
update.
You will get the text with the HTML node:

change 
```
if (editorRef.current !== undefined) {
    if (editorRef.current !== null) {
      const latestEditorState = editorRef.current.getEditorState();
      const textContent = latestEditorState.read(() =>
        //You could change getTextContent() for your purpose
        $getRoot().getTextContent()
      );
      console.log(textContent);
    }
  }
```
to
```
if (editorRef.current !== undefined) {
        if (editorRef.current !== null) {            
            register("content",{value: JSON.stringify(editorRef.current.getEditorState())})
        }
    }
```

**Reply by PunkFleet** [reply]
```
            register("content",{value: JSON.stringify(editorRef.current.getEditorState())})
```
This code is not working now on lexical v0.8.0

**Comment by gabrielbarbosa1207** [comment]
Hi guys, I'm trying to create something pretty similar to the previous example. However, when I'm trying to save the text editor content in JSON string or HTML, it saves as JSON Format. I've tried to use this code from Lexical documentation:

import {$generateHtmlFromNodes} from '@lexical/html';

const htmlString = $generateHtmlFromNodes(editor, selection | null);

But I'm still unable to save it as I wanna.

This is the relevant code section:

  ```
const editorRef: any = React.useRef();

  const onSubmit = async (data: any) => {
  
    // Get the current editor state from the editor ref.
    const editorState = editorRef.current.getEditorState();
  
    // Convert the editor state to a JSON string.
    const content = JSON.stringify(editorState, null, 2);
  
    // Set the content of the data object.
    data.content = content;
  
    console.log("Data published: ", data.content, data);
    try {
      const response = await axios.post('http://localhost:80/create', data);
      console.log(response.data);
    } catch (error) {
      console.error(error);
    }
  };
```

**Comment by moy2010** [comment]
The editor state has a `toJSON` method

**Reply by gabrielbarbosa1207** [reply]
Do you know how can I convert it to HTML before saving it?

**Reply by moy2010** [reply]
Sure: https://lexical.dev/docs/concepts/serialization

**Reply by gabrielbarbosa1207** [reply]
I'm using this code to convert it to HTML. It doesn't seem to work well   

 // Get the current editor state from the editor ref.
    const editorState = editorRef.current.getEditorState();
  
   // Convert the editor state to html
    const htmlString = $generateHtmlFromNodes(editorState, null);  
    
// Set the content of the data object.
    data.content = htmlString;

**Reply by moy2010** [reply]
You have to do it like in the examples from the documentation.

**Reply by gabrielbarbosa1207** [reply]
When I'm implementing `$generateHtmlFromNodes(editorState, null);`. I'm always getting this error -> "Unable to find an active editor state. State helpers or node methods can only be used synchronously during the callback of editor.update() or editorState.read()".

**Reply by abelsj60** [reply]
The reason for the error is that many Lexical functions assume that you will be invoking them from within a Lexical closure. 

If you don't, they won't be able to access the editor's nodes in order to perform the sort of operations you want here. 

The reason you haven't noticed is that you are using a `ref` to access the editorState, so you are probably invoking it a long way from your `LexicalComposer`. You sometimes get a return value because `getEditorState()` sits on the editor instance itself, so it's accessible. 

No idea if `getEditorState()` is accessing a closure internally or not. 

You should be able to take another step forward if you slightly restructure your code around the closure: 

```ts
const htmlEditorState = editorRef.current.getEditorState().read(() => {
   // ...your code goes here. you're in a closure, so you should be able to do everything Lexical lets you do.
   return result; 
}
// you should now be able to send the HTML string you want...
```

Hope that helps. -j

**Reply by moy2010** [reply]
@abelsj60 , the issue is much simpler. `$generateHtmlFromNodes` takes an instance of the editor, not of the editor state.

**Reply by abelsj60** [reply]
You can't see it, but I am now roaring with laughter. I got it wrong! lol

**Reply by gabrielbarbosa1207** [reply]
@moy2010  @abelsj60   You guys mean something like this? -> 

```
const editorRef: any = React.useRef();

  const onSubmit = async (data: any) => {
    alert("Event trigged");
  
    const htmlString = $generateHtmlFromNodes(editorRef, null);
  
    // Set the content of the data object.
    data.content = htmlString;
  
    console.log("Data published: ", data.content, data)
```

**Reply by gabrielbarbosa1207** [reply]
Not sure if it's the best practice, but it finally worked for me:

```
 editorRef.current.update(() => {
      // Get the current editor state.
      const editorState = editorRef.current.getEditorState();
  
      // Convert the editor state to a JSON string.
     const content = JSON.stringify(editorState, null, 2);
  
      // Generate HTML from the editor state.
     const htmlString = $generateHtmlFromNodes(editorRef.current, null);
  
      // Log the generated content.
      console.log('JSON content:', content);
      console.log('HTML content:', htmlString);
  
      // Set the content of the data object.
      data.content = content;
    });
```

**Reply by abelsj60** [reply]
It doesn't look bad, at first blush. You seem to have covered all concerns admirably. Congrats.
